### DISCUSSION TITLE
The supported cmake version for buildsystems/cmake/vcpkg.cmake is 3.7.2

### DISCUSSION THREAD

**Discussion by BillyONeal** [microsoft/vcpkg#19048] [General]
I've been doing some testing in order to get ready to ship binary releases of the vcpkg tool for Linux and MacOS, to characterize how we want to make sure everything works, etc. I started with the .NET Core support matrix, and evaluated what was necessary to get vcpkg working with all that. The table is as follows:

![image](https://user-images.githubusercontent.com/1544943/126567201-d4b56c71-16fe-41c6-9a79-0d05f0adae54.png)

I don't think supporting CMake 2.x is practical for the project at this time, and the oldest cmake bundled with a distro that we may care about seems to be CMake 3.5.1 on Ubuntu 16.04 LTS.

Thoughts?

**Comment by autoantwort** [comment]
If you don't know it: https://gitlab.kitware.com/cmake/community/-/wikis/CMake-Versions-on-Linux-Distros

On mac is it always the newest version through `brew`

**Reply by BillyONeal** [reply]
It's true that that's available through brew, but it would be nice to get to a place that we don't need to be bootstrapped through brew.

**Comment by dg0yt** [comment]
For CMake, see https://repology.org/project/cmake/versions. And there is a badge:
[![Packaging status](https://repology.org/badge/vertical-allrepos/cmake.svg)](https://repology.org/project/cmake/versions)

**Comment by Neumann-A** [comment]
Even if vcpkg says it supports 3.5.1 this isn't really true due to changes to `Find<PackageNames>.cmake ` which might break wrappers. 

Considering that CMake is available via snap, pip and apt if configured (https://apt.kitware.com/), I don't see why vcpkg should support such old CMake versions. (Especially since the entry with 3.5.1 seems not to even work today and needs apt reconfiguration any way.....)

Having said that everything > ~2 years seems dropable as such Ubuntu 18.04 is for me the baseline which would be cmake 3.10.2 (but for me the real baseline would be more around 3.15-3.17, because ports like qt6 have `cmake_minimum_required(VERSION 3.14...3.19)` in the **installed** `-Config.cmake`)

**Reply by BillyONeal** [reply]
> Even if vcpkg says it supports 3.5.1 this isn't really true due to changes to `Find<PackageNames>.cmake ` which might break wrappers.

It seems like it's our toolchain's job to fix that up if it is a library we're deploying?

> Considering that CMake is available via snap, pip and apt if configured (https://apt.kitware.com/), I don't see why vcpkg should support such old CMake versions. (Especially since the entry with 3.5.1 seems not to even work today and needs apt reconfiguration any way.....)

It doesn't work today because of the `std::filesystem` dependency to build vcpkg forcing use of gcc >= 7, which I am currently addressing explicitly so that we can target these older linuxes.

> Having said that everything > ~2 years seems dropable as such Ubuntu 18.04 is for me the baseline which would be cmake 3.10.2 (but for me the real baseline would be more around 3.15-3.17, because ports like qt6 have `cmake_minimum_required(VERSION 3.14...3.19)` in the **installed** `-Config.cmake`)

This is talking about vcpkg itself sets as a floor -- individual ports are free to require later versions if they wish.

**Reply by Neumann-A** [reply]
> It seems like it's our toolchain's job to fix that up if it is a library we're deploying?

@BillyONeal: The title mentions `buildsystems/cmake/vcpkg.cmake` so  I assumed its not just the vcpkg binary itself.

**Reply by BillyONeal** [reply]
Yes, this is about that file and only that file. For portfile.cmake and friends, vcpkg already downloads a current copy of cmake to use. For SomePort-Config.cmake, individual ports can decide which versions of cmake they care about. buildsystems/cmake/vcpkg.cmake is special because (1) it runs in the end user's build rather than in vcpkg so it can't assume our recently downloaded cmake, and (2) no port can support an older copy of cmake than that supports, even if it wishes to.

We've had requests from, for example, the Azure and AWS SDK folks. They would like to use vcpkg as their primary distribution mechanism, but have enterprise-y customers locked to said old linuxes. The desire is to enable those port authors to help those customers if they wish, not to force qt6 to support those customers if it does not.

**Reply by dg0yt** [reply]
It is a strange dissonance for manifest mode users when the minimum version of cmake varies by wrapper (or by contributor).

It is a strange dissonance for contributors when portfiles can use the latest cmake features, but (some) wrappers shall use the state of cmake from five years ago.

How do the enterprise consumers deal with current situation which in fact needs CMake 3.15 (CMake 3.4 when #18898 gets merged eventually)?

If the minimum version of cmake varies by wrapper, shouldn't it become an explicit dependency? So that the user project knows problems in advance, before building all of Qt 6, gdal, etc.? It would also make cmake requirement changes more obvious in code review.

**Reply by BillyONeal** [reply]
>It is a strange dissonance for manifest mode users when the minimum version of cmake varies by wrapper (or by contributor).

I don't see any intersection with manifests here. If you add a thing that wants a recent cmake to your manifest, you need to be using a recent enough cmake. If you don't add such a thing, you don't need such a recent cmake. This is no different than, for example, adding something that wants CUDA to your manifest where you need CUDA to be present before things will work.

>How do the enterprise consumers deal with current situation which in fact needs CMake 3.15 (CMake 3.4 when #18898 gets merged eventually)?

They don't deal with the current situation. That is, in fact, the problem. That review is what kicked off this discussion: that we need tests for our documented minimum requirements, prompting a reevaluation of what those minimum requirements should be. Clearly, the previous 3.1 policy was not substantiated by actual customer needs. (And was not tested)

>If the minimum version of cmake varies by wrapper, shouldn't it become an explicit dependency? So that the user project knows problems in advance, before building all of Qt 6, gdal, etc.? It would also make cmake requirement changes more obvious in code review.

When vcpkg is running, in general, we don't know how the results will be used. Perhaps manifest mode could make the assumption that the installed tree it produces is unlikely to be used outside of the calling cmake script and warn.

**Reply by BillyONeal** [reply]
> > If the minimum version of cmake varies by wrapper, shouldn't it become an explicit dependency? So that the user project knows problems in advance, before building all of Qt 6, gdal, etc.? It would also make cmake requirement changes more obvious in code review.
> 
> When vcpkg is running, in general, we don't know how the results will be used. Perhaps manifest mode could make the assumption that the installed tree it produces is unlikely to be used outside of the calling cmake script and warn.

It should also be noted that the qt6 build they did is not entirely a waste because rerunning their configure after downloading a current cmake should be able to reuse the bits :)

**Reply by dg0yt** [reply]
> If you add a thing that wants a recent cmake to your manifest, you need to be using a recent enough cmake. If you don't add such a thing, you don't need such a recent cmake.

The problem is not adding ports, but ports being updated, including dependencies being added.
(What if the proposed zlib wrapper would need CMake 3.21?)

> When vcpkg is running, in general, we don't know how the results will be used.

This is okay. But as a user, I want to understand the impact of relying on ports from the vcpkg registry. Even if I can help with contributions, I do need to rely on some firm grounds.

> It should also be noted that the qt6 build they did is not entirely a waste because rerunning their configure after downloading a current cmake should be able to reuse the bits :)

The concern is not the time to rebuild Qt. The concern is that I tested successfully with my version of CMake today, and the tiniest wrapper my project loads today will suddenly require a new version of CMake.

**Reply by dg0yt** [reply]
To draw the picture differently:
Ports are built in a "latest CMake" world.
The toolchain is used by users in a "legacy CMake" world.
Wrappers are published from the "latest CMake" world for use in the "legacy CMake" world.
I'm concerned about control of new CMake requirements slipping into the  "legacy CMake" world via the wrappers.

**Reply by Neumann-A** [reply]
To add to this:
It is especially problematic if "latest CMake" returns something different via find_library|find_package then  "legacy CMake" world. As such vcpkg should always recommend the CMake version used within vcpkg.

**Reply by BillyONeal** [reply]
>The problem is not adding ports, but ports being updated, including dependencies being added.
(What if the proposed zlib wrapper would need CMake 3.21?)
> [...]
>The concern is not the time to rebuild Qt. The concern is that I tested successfully with my version of CMake today, and the tiniest wrapper my project loads today will suddenly require a new version of CMake.

You would have to change the version of zlib or qt you're referencing (or the baseline SHA) for that to happen.

>This is okay. But as a user, I want to understand the impact of relying on ports from the vcpkg registry. Even if I can help with contributions, I do need to rely on some firm grounds.

I don't think this is really relevant to the cmake version supported discussion. Again, this is no different vs. today with CUDA or other ports that require system bits to be present.

> To draw the picture differently:
> Ports are built in a "latest CMake" world.
> The toolchain is used by users in a "legacy CMake" world.
> Wrappers are published from the "latest CMake" world for use in the "legacy CMake" world.
> I'm concerned about control of new CMake requirements slipping into the "legacy CMake" world via the wrappers.

There is nothing vcpkg can do about this. We don't own the caller of buildsystems\cmake\vcpkg.cmake's cmake copy. We could intentionally fail older copies, but that doesn't seem to be helpful to anyone.

>It is especially problematic if "latest CMake" returns something different via find_library|find_package then "legacy CMake" world. As such vcpkg should always recommend the CMake version used within vcpkg.

That is our recommendation, but we support things outside of our recommendation. We don't recommend x64-windows-static-md either but lots of people are using it :)

**Reply by dg0yt** [reply]
@BillyONeal wrote:
> ... that we need tests for our documented minimum requirements, prompting a reevaluation of what those minimum requirements should be. Clearly, the previous 3.1 policy was not substantiated by actual customer needs. (And was not tested)

#19034 is such a test. It is an elaborated variant of what lead to #18898.

**Comment by dg0yt** [comment]
@BillyONeal Did you come to any decision? I would update #19034 when 3.5.1 becomes the new standard.

**Comment by BillyONeal** [comment]
Changed title to 3.7.2 because I don't think we should care about Ubuntu 16.04 any longer.

**Reply by dg0yt** [reply]
Now we only need a validation facility.

**Comment by sorokin** [comment]
More than 3 years passed since the last decision was made. I wonder if vcpkg should update its cmake version requirements.

I compiled [a little table](https://github.com/user-attachments/files/20432718/cmake-versions.txt) listing distributions, cmake versions and their EOL dates.
<img src="https://github.com/user-attachments/assets/de4c4871-5931-40a9-8b8e-d3c300429ff0" width="50%">

Ubuntu 20.04 EOLs in 4 days. Perhaps requiring minimum cmake 3.13 or 3.16 is practical now?

**Reply by dg0yt** [reply]
The current minimum version is 3.16.

**Reply by dg0yt** [reply]
BTW the `cmake-user` port (in `scripts/test_ports`) tests some standard CMake find modules with the minimum version.

**Reply by sorokin** [reply]
> The current minimum version is 3.16.

Oh, thank you! Good to know.

I tried googling the minimal required cmake version, but the closest match google shows is the link to here. Now I see that this requirement is even documented [vcpkg/concepts/supported-hosts](https://learn.microsoft.com/en-us/vcpkg/concepts/supported-hosts).

So, sorry for bothering you and thank you for the reply.

**Reply by dg0yt** [reply]
No problem, and thanks for reporting the results of the second check.
